stacktrace.go

```diff
@@ -20,23 +20,72 @@
 
 package zap
 
-import "runtime"
+import (
+	"runtime"
+	"strings"
+	"sync"
+
+	"go.uber.org/zap/internal/bufferpool"
+)
+
+var (
+	defaultUintptrSliceSize = 128
+	uintptrPool             = sync.Pool{
+		New: func() interface{} {
+			return make([]uintptr, defaultUintptrSliceSize)
+		},
+	}
+)
 
 // takeStacktrace attempts to use the provided byte slice to take a stacktrace.
 // If the provided slice isn't large enough, takeStacktrace will allocate
 // successively larger slices until it can capture the whole stack.
-func takeStacktrace(buf []byte, includeAllGoroutines bool) string {
-	if len(buf) == 0 {
-		// We may have been passed a nil byte slice.
-		buf = make([]byte, 1024)
+func takeStacktrace() string {
+	buffer := bufferpool.Get()
+	programCounters := getUintptrSlice()
+	n := runtime.Callers(0, programCounters)
+	size := defaultUintptrSliceSize
+	for n >= size {
+		size *= 2
+		// Do not put programCounters back in pool, will put larger buffer in
+		// This is in case our size is always wrong, we optimize to pul
+		// correctly-sized buffers back in the pool
+		programCounters = make([]uintptr, size)
+		n = runtime.Callers(0, programCounters)
 	}
-	n := runtime.Stack(buf, includeAllGoroutines)
-	for n >= len(buf) {
-		// Buffer wasn't large enough, allocate a larger one. No need to copy
-		// previous buffer's contents.
-		size := 2 * n
-		buf = make([]byte, size)
-		n = runtime.Stack(buf, includeAllGoroutines)
+	programCounters = programCounters[:n]
+
+	i := 0
```

Comment:
We can't, because of how we skip as of this PR - this variable is used for whether to add a newline or not